### PR TITLE
[FLINK-35625][cli] Merge "flink run" and "flink run-application" functionality, deprecate "run-application"

### PR DESCRIPTION
**Pull request by ferenc-csaky**
## What is the purpose of the change

Merge the functionality of "flink run" and "flink run-application" under the "run" command to make job deployment more consistent. Also deprecate "run-application" so it can be removed together with "yarn-per-job" in the future.

## Brief change log

- Introduced `isApplication(...)` check method in `FrontendCli` to determine necessary deployment steps.
- Handle application mode deployment in `FrontendCli#run`.
- In `FrontendCli#runApplication`, simply call `FrontendCli#run` after argument check.
- Added unit tests to cover the introduced code flow changes in `CliFrontendRunTests`.
- Updated, adapted CLI help/usage messages, and relevant config option documentation.
- Marked `run-application` deprecated in code, config docs, and help/usage messages. 

## Verifying this change

Unit tests were added to cover the introduced changes.

Also did a manual check, where I executed the following on a locally built Flink and `minikube` setup:
```sh
$ ./bin/flink run-application \
    --target kubernetes-application \
    -Dkubernetes.cluster-id=flink-app-cluster-1 \
    -Dkubernetes.container.image.ref=flink:1.19 \
    local:///opt/flink/examples/streaming/TopSpeedWindowing.jar

$ ./bin/flink run \
    --target kubernetes-application \
    -Dkubernetes.cluster-id=flink-app-cluster-2 \
    -Dkubernetes.container.image.ref=flink:1.19 \
    local:///opt/flink/examples/streaming/TopSpeedWindowing.jar

$ kubectl get pods
NAME                                         READY   STATUS    RESTARTS   AGE
flink-app-cluster-1-778bb58994-q8hg5         1/1     Running   0          30s
flink-app-cluster-1-taskmanager-1-1          1/1     Running   0          22s
flink-app-cluster-2-6555f5fc76-2sspj         1/1     Running   0          18s
flink-app-cluster-2-taskmanager-1-1          1/1     Running   0          9s
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java [hunk 1]
```diff
@@ -234,20 +205,77 @@ protected void run(String[] args) throws Exception {
...
+
+            try (PackagedProgram program =
+                    getPackagedProgram(programOptions, effectiveConfiguration)) {
+                executeProgram(effectiveConfiguration, program);
+            }
         }
     }
 
+    /** Checks if the given job should be deployed in Application Mode. */
+    protected boolean isApplication(
```
Comment (gaborgsomogyi) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:260]: Maybe `isDeploymentTargetApplication` or something which is more meaningful?

File: flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendRunTest.java [hunk 2]
```diff
@@ -285,7 +345,7 @@ public static void verifyCliFrontend(
         RunTestingCliFrontend testFrontend =
                 new RunTestingCliFrontend(
                         configuration,
-                        clusterClientServiceLoader,
+                        new DefaultClusterClientServiceLoader(),
```
Comment (gaborgsomogyi) [inline flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendRunTest.java:348]: Not sure why this change is needed? I see `new DefaultClusterClientServiceLoader` other places too, is it possible to do it once per CLI?

**Review by gaborgsomogyi — COMMENTED** [review]
Thanks, only some minors found.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java (hunk 1 above)
Comment (gaborgsomogyi) [reply flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontend.java:260]: With meaningful name comment can be deleted :)

File: flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendRunTest.java (hunk 2 above)
Comment (ferenc-csaky) [reply flink-clients/src/test/java/org/apache/flink/client/cli/CliFrontendRunTest.java:348]: This was like this before, I just removed one unused `verifyCliFrontend` method. Anyways, this can be moved to a static constant of the `RunTestingCliFrontend` nested class.

**Review by gaborgsomogyi — APPROVED** [review]

**Comment by gaborgsomogyi** [comment]
Unless further comments are not coming I'm intended to merge it beginning of next week.

**Comment by ferenc-csaky** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
@flinkbot run azure

**Comment by gaborgsomogyi** [comment]
Waiting on azure and in case of success merging...
